GitCommands/Git/GitCommand.cs

```diff
@@ -49,5 +49,13 @@ public override string ToString()
             return ToLine();
         }
 
+        /// <summary>
+        /// Validates if the supplied arguments are correct
+        /// </summary>
+        public virtual void Validate()
```

Comment:
Is there any reason you didn't do this?

```
    public virtual string ToLine()
    {
        Validate();
        return GitComandName() + " " + CollectArguments().Join(" ");
    }
```

Right now, the validation is responsibility of a caller, which isn't obvious to the caller nor enforced in anyway.
Should the caller to forget to validate, it could lead to a fatal app crash....


I am unsure if we are gaining anything by introducing `Validate()` at all. It adds a disconnect between the validation and the argument enumeration. It also splits responsibility of who and when should validate.
Was there any reason you could't validate argument in `CollectArguments()`?

Reply:
Yes, you are right Validate should be inside the body of ToLine.

>Was there any reason you could't validate argument in CollectArguments()?

Because I wanted to split the validation and collecting logics. There is no sense to collect arguments if they are invalid. CollectArguments returns IEnumerable, you could call CollectArguments().Any() and don't get any exception. If collecting arguments involved some resource consuming operations, it is better to perform the whole validation process first.
This is a part of my understanding of SRP.

Reply:
> CollectArguments returns IEnumerable, you could call CollectArguments().Any() and don't get any exception

![image](https://user-images.githubusercontent.com/4403806/30058705-78df0d72-927f-11e7-82f3-4f58d72f77e5.png)

> This is a part of my understanding of SRP.

SRP talks about classes not methods.

> a class should have only a single responsibility (i.e. changes to only one part of the software's specification should be able to affect the specification of the class)

Reply:
:) This way you have to do the validation upfront (how much the disconnection was decreased?). It still deserves to put the validation to a separate method.